modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4ChunkedEncodingIT.java

```diff
@@ -112,66 +163,110 @@ public Collection<RestHandler> getRestHandlers(
-                    );
-                }
-            }, new BaseRestHandler() {
-                @Override
-                public String getName() {
-                    return EMPTY_ROUTE;
-                }
+                        );
+                    }
+                },
 
-                @Override
-                public List<Route> routes() {
-                    return List.of(new Route(GET, EMPTY_ROUTE));
-                }
+                // only a few random empty chunks
+                new BaseRestHandler() {
+                    @Override
+                    public String getName() {
+                        return EMPTY_ROUTE;
+                    }
+
+                    @Override
+                    public List<Route> routes() {
+                        return List.of(new Route(GET, EMPTY_ROUTE));
+                    }
+
+                    @Override
+                    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) {
+                        return channel -> sendChunksResponse(channel, emptyChunks());
+                    }
+                },
+
+                // keeps on emitting chunks until cancelled
+                new BaseRestHandler() {
+                    @Override
+                    public String getName() {
+                        return INFINITE_ROUTE;
+                    }
 
-                @Override
-                protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) {
-                    return channel -> sendChunksResponse(channel, emptyChunks());
+                    @Override
+                    public List<Route> routes() {
+                        return List.of(new Route(GET, INFINITE_ROUTE));
+                    }
+
+                    @Override
+                    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) {
+                        return channel -> sendChunksResponse(channel, new Iterator<>() {
+                            private static final BytesReference CHUNK = new BytesArray("CHUNK\n");
+
+                            @Override
+                            public boolean hasNext() {
+                                return true;
+                            }
+
+                            @Override
+                            public BytesReference next() {
+                                return CHUNK;
+                            }
+                        });
+                    }
                 }
-            });
+            );
         }
 
         private static void sendChunksResponse(RestChannel channel, Iterator<BytesReference> chunkIterator) {
-            channel.sendResponse(RestResponse.chunked(RestStatus.OK, new ChunkedRestResponseBody() {
-                @Override
-                public boolean isDone() {
-                    return chunkIterator.hasNext() == false;
-                }
+            if (refs.tryIncRef()) {
```

Comment:
Why is not `refs.mustIncRef()`? It seems something has already gone wrong if reference cannot be increased here?

Reply:
I don't think so, for the cancellation test we can hit the end of the `try (withResourceTracker()) {...}` block before the node even starts to handle the request.

Reply:
Ah ok. Thanks for explaining this. In that case, I think we should have a `null` check for `refs` as well.